### ISSUE TITLE
Strange segfault when taking gradient

### ISSUE THREAD

**Issue by yebai** [TuringLang/ThArrays.jl#8]
```julia

julia> using ThArrays

julia> x = Tensor(rand(1,5000), requires_grad=true);

julia> f(x) = begin 
       y=Tensor(0.0, requires_grad=true)
       for i=1:length(x)
          y+=x[i];
       end
       y
       end
f (generic function with 1 method)

julia> y = f(x)
PyTorch.Tensor{Float64, 0}:
2492.13
[ CPUDoubleType{} ]

julia> backward(y)

julia> grad(x)
libc++abi.dylib: terminating with uncaught exception of type c10::Error: dim() called on undefined Tensor (dim at ../c10/core/UndefinedTensorImpl.cpp:24)


signal (6): Abort trap: 6
in expression starting at REPL[6]:1
__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
Allocations: 4838242 (Pool: 4837414; Big: 828); GC: 9
zsh: abort      julia

```

**Comment by KDr2** [comment]
Fixed by https://github.com/TuringLang/ThArrays.jl/commit/5014e73bd1f44ccda0029fc3b0eaa22444dc2a6c.

Before this commit, `eltype(Tensor{T}) == T`, so `x[i]` returns a bare number which is not tracked, so the computional info is lost.

After the commit, `eltype(Tensor{T}) == Tensor{T, 0}` and `x[i]` became a 0-dim Tensor which brings the computional info along.

One can use `t[]` to get the bare number inside `t` which is a 1-element Tensot, e.g., a 0-dim Tensor.
